src/control/server/storage/scm/ndctl.go

```diff
@@ -0,0 +1,277 @@
+//
+// (C) Copyright 2022-2023 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package scm
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/dustin/go-humanize"
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/server/storage"
+)
+
+const (
+	ndctlName         = "ndctl"
+	ndctlRegionType   = "pmem"
+	ndctlRegionDomain = "memory_controller"
+)
+
+var (
+	nsMajMinRegex = regexp.MustCompile(`namespace([0-9]).([0-9])`)
+
+	// Cmd structs for ndctl commandline calls.
+
+	cmdCreateNamespace = pmemCmd{
+		BinaryName: ndctlName,
+		Args:       []string{"create-namespace"},
+	}
+	// expects device name param
+	cmdDisableNamespace = pmemCmd{
+		BinaryName: ndctlName,
+		Args:       []string{"disable-namespace"},
+	}
+	// expects device name param
+	cmdDestroyNamespace = pmemCmd{
+		BinaryName: ndctlName,
+		Args:       []string{"destroy-namespace"},
+	}
+	// returns ns info in json
+	cmdListNamespaces = pmemCmd{
+		BinaryName: ndctlName,
+		Args:       []string{"list", "-N", "-v"},
+	}
+	// returns region info in json
+	cmdListNdctlRegions = pmemCmd{
+		BinaryName: ndctlName,
+		Args:       []string{"list", "-R", "-v"},
+	}
+)
+
+func (cr *cmdRunner) checkNdctl() (errOut error) {
+	cr.checkOnce.Do(func() {
+		if _, err := cr.lookPath("ndctl"); err != nil {
+			errOut = FaultMissingNdctl
+		}
+	})
+
+	return
+}
+
+// For each region, create <nrNsPerSocket> namespaces. Return slice indicating which NUMA nodes the
+// name spaces were created on.
+func (cr *cmdRunner) createNamespaces(regionPerSocket socketRegionMap, nrNsPerSock uint) ([]int, error) {
+	if nrNsPerSock < minNrNssPerSocket || nrNsPerSock > maxNrNssPerSocket {
+		return nil, errors.Errorf("unexpected number of namespaces requested per socket: want [%d-%d], got %d",
+			minNrNssPerSocket, maxNrNssPerSocket, nrNsPerSock)
+	}
+
+	// For the moment assume 1:1 mapping of sockets to regions
+	sockIDs := regionPerSocket.keys()
+	if len(sockIDs) == 0 {
+		return nil, errors.New("expected non-zero number of pmem regions in input map")
+	}
+	cr.log.Debugf("creating %d namespaces on each of the following socket(s): %v", nrNsPerSock,
+		sockIDs)
+
+	// As the selector is socket, look up the ndctl region with the same ISetID as the ipmctl
+	// region with specified socket ID. This may not work when the ISetID overflows in ndctl
+	// output. If all sockIDs can be mapped to existent entries in ndctl regions by isetid
```

Comment:
What are the possible cases where this overflow occurs? Is this a bug we need to report to ndctl?

Reply:
It is a bug and there are various related tickets regarding overflows in output values, they look like they have been resolved in more recent versions (the distro versions are relatively old in relationship to recent releases). I didn't put a ticket number in the comment because there isn't a direct match.

Reply:
The mapping between regions using the ISetID is an improvement but the fallback method of using the socket ID as the region index identifier should work fine in the vast majority of cases.